### ISSUE TITLE
[OpenCL-am57xx] SDK 6.2.0.81

### ISSUE THREAD

**Issue by RobertCNelson** [beagleboard/Latest-Images#11]
OpenCL: resync with TI's SDK

http://arago-project.org/git/projects/oe-layersetup.git?p=projects/oe-layersetup.git;a=commit;h=d5be8dfa31b41cdc6fd0fa187f91d4a93eb52b43

Software Versions:

ti-llvm-3.6: ti-llvm3.6-3.6-r7
ti_cgt_c6000_8.3.2
cmem: cmem-4.16.00.00+gitAUTOINC+4f970f053b-r0:
kernel: linux-ti-staging-4.19.79+gitAUTOINC+77dfab56c6-r7a.arago5.tisdk1.9
ti-ipc: ti-ipc-3.50.04.08-r0.2
opencl: opencl-1.2.00.02-r0.0
tidl-api: 1.4.0-r0.1

Stage 1: Build TI SDK (jenkins): (Master directions: http://software-dl.ti.com/processor-sdk-linux/esd/docs/latest/linux/Overview_Building_the_SDK.html )

```
#!/bin/bash -x

pwd

release="06.02.00.81"

if [ -d tisdk ] ; then
   rm -rf ./tisdk/ || true
fi

wget -c https://developer.arm.com/-/media/Files/downloads/gnu-a/8.3-2019.03/binrel/gcc-arm-8.3-2019.03-x86_64-arm-linux-gnueabihf.tar.xz
wget -c https://developer.arm.com/-/media/Files/downloads/gnu-a/8.3-2019.03/binrel/gcc-arm-8.3-2019.03-x86_64-aarch64-linux-gnu.tar.xz

if [ ! -d ./gcc-arm-8.3-2019.03-x86_64-arm-linux-gnueabihf ] ; then
    rm -rf ./gcc-arm-8.3-2019.03-x86_64-arm-linux-gnueabihf || true
fi
tar xf gcc-arm-8.3-2019.03-x86_64-arm-linux-gnueabihf.tar.xz

if [ ! -d ./gcc-arm-8.3-2019.03-x86_64-aarch64-linux-gnu ] ; then
    rm -rf ./gcc-arm-8.3-2019.03-x86_64-aarch64-linux-gnu || true
fi
tar xf gcc-arm-8.3-2019.03-x86_64-aarch64-linux-gnu.tar.xz

export TOOLCHAIN_PATH_ARMV7=/var/lib/jenkins/workspace/ti-sdk-${release}/gcc-arm-8.3-2019.03-x86_64-arm-linux-gnueabihf
export TOOLCHAIN_PATH_ARMV8=/var/lib/jenkins/workspace/ti-sdk-${release}/gcc-arm-8.3-2019.03-x86_64-aarch64-linux-gnu

${TOOLCHAIN_PATH_ARMV7}/bin/arm-linux-gnueabihf-gcc -v
${TOOLCHAIN_PATH_ARMV8}/bin/aarch64-linux-gnu-gcc -v

git clone git://arago-project.org/git/projects/oe-layersetup.git tisdk

if [ -d tisdk ] ; then
    cd tisdk/
    
    ./oe-layertool-setup.sh -f configs/processor-sdk/processor-sdk-${release}-config.txt
    
    mkdir -p /var/lib/jenkins/workspace/ti-sdk-${release}/tisdk/downloads/
    if [ -d /var/lib/jenkins/dl_ti-sdk-${release}/ ] ; then
        rsync -av /var/lib/jenkins/dl_ti-sdk-${release}/ /var/lib/jenkins/workspace/ti-sdk-${release}/tisdk/downloads/
    fi

    cd ./build

    . conf/setenv
    export TOOLCHAIN_PATH_ARMV7=/var/lib/jenkins/workspace/ti-sdk-${release}/gcc-arm-8.3-2019.03-x86_64-arm-linux-gnueabihf
    export TOOLCHAIN_PATH_ARMV8=/var/lib/jenkins/workspace/ti-sdk-${release}/gcc-arm-8.3-2019.03-x86_64-aarch64-linux-gnu

    MACHINE=am57xx-evm bitbake opencl-monitor ipumm-fw vpdma-fw cmem-mod tidl-api
    mkdir -p /var/lib/jenkins/dl_ti-sdk-${release}/
    rsync -av /var/lib/jenkins/workspace/ti-sdk-${release}/tisdk/downloads/ /var/lib/jenkins/dl_ti-sdk-${release}/
fi
```

SDK Firmware:
opencl-monitor: https://github.com/rcn-ee/sdk-firmware/tree/master/opencl-monitor/1.2.00.02-r0.0/files
opencl-tidl-fw: https://github.com/rcn-ee/sdk-firmware/tree/master/opencl-tidl-fw/01.02.00.01-r0/files
opencl-monitor-ipu: https://github.com/rcn-ee/sdk-firmware/tree/master/opencl-monitor-ipu/1.2.00.02-r0.0/files

meta-arago: ti2019.05

http://arago-project.org/git/?p=meta-arago.git;a=shortlog;h=refs/heads/ti2019.05

llvm-3.6:
http://arago-project.org/git/?p=meta-arago.git;a=tree;f=meta-arago-extras/recipes-core/llvm;h=59cb30cdb4c5375ce473521a9dac1958a3768d9a;hb=refs/heads/ti2019.05

Package Builds:
ti-llvm-3.6 (3.6-git20190111.1)
ti-c6000-cgt-v8.3 (8.3.2-0rcnee1)
libtiulm-dev_1.4.0-git20160915-0rcnee3
ti-ipc-dra7xx_3.50.04.08-rcnee0
cmem_4.16.00.00-git20181220.1
gdbserver-c6x (1.1.0-git20190509.0)
firmware-am57xx-opencl-monitor (1.2.00.02-r0.0rcnee0)
firmware-am57xx-opencl-tidl (01.02.00.01-r0rcnee0)

ToDo:
Port:
https://github.com/rcn-ee/ti-opencl/commits/v01.01.17.01-bb.org
to:
https://github.com/rcn-ee/ti-opencl/commits/v01.02.00.02-bb.org

and rebuild...

Edit Old notes:
https://gist.github.com/RobertCNelson/aca2d52e962c80343857

```
http://software-dl.ti.com/processor-sdk-linux/esd/AM57X/06_01_00_08/index_FDS.html
http://software-dl.ti.com/processor-sdk-linux/esd/AM57X/06_01_00_08/exports/software_manifest.htm
1.1.19.02-r0.0.49 4.19.59

http://software-dl.ti.com/processor-sdk-linux/esd/AM57X/06_00_00_07/index_FDS.html
http://software-dl.ti.com/processor-sdk-linux/esd/AM57X/06_00_00_07/exports/software_manifest.htm
1.1.18.00-r0.1.0 4.19.38

http://software-dl.ti.com/processor-sdk-linux/esd/AM57X/05_03_00_07/index_FDS.html
http://software-dl.ti.com/processor-sdk-linux/esd/AM57X/05_03_00_07/exports/software_manifest.htm
1.1.18.00-r0.0.0 4.14.79-r9.0
http://software-dl.ti.com/processor-sdk-linux/esd/docs/05_03_00_07/linux/Overview_Building_the_SDK.html

http://software-dl.ti.com/processor-sdk-linux/esd/AM57X/05_02_00_10/index_FDS.html
http://software-dl.ti.com/processor-sdk-linux/esd/AM57X/05_02_00_10/exports/software_manifest.htm
1.1.17.01-r1.0.0 4.14.79
http://software-dl.ti.com/processor-sdk-linux/esd/docs/05_02_00_10/linux/Overview_Building_the_SDK.html#introduction
```

**Comment by RobertCNelson** [comment]
Helper for when porting the library...
```
#!/bin/bash

git_clone_checkout () {
	if [ -d ./${name} ] ; then
		rm -rf ./${name} || true
	fi

	git clone git@github.com:rcn-ee/ti-opencl.git ${name}
	cd ./${name}/
	git checkout ${git_sha} -b tmp
	cd ../
}

git_clone_branch () {
	if [ -d ./${branch} ] ; then
		rm -rf ./${branch} || true
	fi
	git clone -b ${branch} git@github.com:rcn-ee/ti-opencl.git ${branch}
}

git_sha="61d9dbae3ee40f8023ec624601c2a0cc9e80220f" ; name="v01.01.17.01" ; git_clone_checkout
git_sha="b1dfed9af21a1236380b9d643c432e4f0c619fce" ; name="v01.02.00.02" ; git_clone_checkout

branch="v01.01.17.01-bb.org" ; git_clone_branch
branch="v01.02.00.02-bb.org" ; git_clone_branch
```

**Comment by RobertCNelson** [comment]
Conversion:
```
mkdir ./host/include/CL/TI/
git mv ./host/include/CL/cl* ./host/include/CL/TI/
git mv ./host/include/CL/opencl.h ./host/include/CL/TI/


find . -type f -exec sed -i 's:<CL/cl.hpp:<CL/TI/cl.hpp:g' {} \;
find . -type f -exec sed -i 's:<CL/cl.h:<CL/TI/cl.h:g' {} \;
find . -type f -exec sed -i 's:<CL/cl_d3d10.h:<CL/TI/cl_d3d10.h:g' {} \;
find . -type f -exec sed -i 's:<CL/cl_d3d11.h:<CL/TI/cl_d3d11.h:g' {} \;
find . -type f -exec sed -i 's:<CL/cl_dx9_media_sharing.h:<CL/TI/cl_dx9_media_sharing.h:g' {} \;
find . -type f -exec sed -i 's:<CL/cl_egl.h:<CL/TI/cl_egl.h:g' {} \;
find . -type f -exec sed -i 's:<CL/cl_ext.h:<CL/TI/cl_ext.h:g' {} \;
find . -type f -exec sed -i 's:<CL/cl_gl_ext.h:<CL/TI/cl_gl_ext.h:g' {} \;
find . -type f -exec sed -i 's:<CL/cl_gl.h:<CL/TI/cl_gl.h:g' {} \;
find . -type f -exec sed -i 's:<CL/cl_platform.h:<CL/TI/cl_platform.h:g' {} \;
find . -type f -exec sed -i 's:<CL/opencl.h:<CL/TI/opencl.h:g' {} \;
```

**Comment by RobertCNelson** [comment]
```
[ 22%] Building CXX object src/CMakeFiles/TIOpenCL.dir/api/api_command.cpp.o
In file included from /usr/include/CL/cl.h:32,
                 from /home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/core/icd.h:34,
                 from /home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/core/commandqueue.h:38,
                 from /home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/api/api_command.cpp:33:
/usr/include/CL/cl_version.h:34:104: note: #pragma message: cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)
 #pragma message("cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)")
                                                                                                        ^
[ 24%] Building CXX object src/CMakeFiles/TIOpenCL.dir/api/api_device.cpp.o
In file included from /usr/include/CL/cl.h:32,
                 from /home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/api/api_device.cpp:34:
/usr/include/CL/cl_version.h:34:104: note: #pragma message: cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)
 #pragma message("cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)")
                                                                                                        ^
[ 25%] Building CXX object src/CMakeFiles/TIOpenCL.dir/api/api_event.cpp.o
[ 26%] Building CXX object src/CMakeFiles/TIOpenCL.dir/api/api_kernel.cpp.o
In file included from /usr/include/CL/cl.h:32,
                 from /home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/api/api_kernel.cpp:33:
/usr/include/CL/cl_version.h:34:104: note: #pragma message: cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)
 #pragma message("cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)")
                                                                                                        ^
[ 27%] Building CXX object src/CMakeFiles/TIOpenCL.dir/api/api_platform.cpp.o
In file included from /usr/include/CL/cl.h:32,
                 from /home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/api/api_platform.cpp:34:
/usr/include/CL/cl_version.h:34:104: note: #pragma message: cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)
 #pragma message("cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)")
                                                                                                        ^
[ 28%] Building CXX object src/CMakeFiles/TIOpenCL.dir/api/api_program.cpp.o
In file included from /usr/include/CL/cl.h:32,
                 from /home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/api/api_program.cpp:33:
/usr/include/CL/cl_version.h:34:104: note: #pragma message: cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)
 #pragma message("cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)")
                                                                                                        ^
[ 29%] Building CXX object src/CMakeFiles/TIOpenCL.dir/api/api_context.cpp.o
[ 31%] Building CXX object src/CMakeFiles/TIOpenCL.dir/api/api_enqueue.cpp.o
[ 32%] Building CXX object src/CMakeFiles/TIOpenCL.dir/api/api_flush.cpp.o
In file included from /usr/include/CL/cl.h:32,
                 from /home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/api/api_flush.cpp:33:
/usr/include/CL/cl_version.h:34:104: note: #pragma message: cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)
 #pragma message("cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)")
                                                                                                        ^
[ 33%] Building CXX object src/CMakeFiles/TIOpenCL.dir/api/api_memory.cpp.o
In file included from /usr/include/CL/cl.h:32,
                 from /home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/api/api_memory.cpp:33:
/usr/include/CL/cl_version.h:34:104: note: #pragma message: cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)
 #pragma message("cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 220 (OpenCL 2.2)")
                                                                                                        ^
/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/api/api_memory.cpp: In function ‘void* __malloc_msmc(size_t)’:
/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/api/api_memory.cpp:533:27: error: ‘CL_MEM_USE_MSMC_TI’ was not declared in this scope
     return clMalloc(size, CL_MEM_USE_MSMC_TI, NULL);
                           ^~~~~~~~~~~~~~~~~~
/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/host/src/api/api_memory.cpp:533:27: note: suggested alternative: ‘CL_MEM_USE_HOST_PTR’
     return clMalloc(size, CL_MEM_USE_MSMC_TI, NULL);
                           ^~~~~~~~~~~~~~~~~~
                           CL_MEM_USE_HOST_PTR
make[5]: *** [src/CMakeFiles/TIOpenCL.dir/build.make:180: src/CMakeFiles/TIOpenCL.dir/api/api_memory.cpp.o] Error 1
make[5]: Leaving directory '/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/build/am57'
make[4]: *** [CMakeFiles/Makefile2:265: src/CMakeFiles/TIOpenCL.dir/all] Error 2
make[4]: Leaving directory '/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/build/am57'
make[3]: *** [Makefile:152: all] Error 2
make[3]: Leaving directory '/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/build/am57'
make[2]: *** [makefile:93: install] Error 2
make[2]: Leaving directory '/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0'
dh_auto_build: make -j1 BUILD_EVE_FIRMWARE=0 BUILD_OUTPUT=lib BUILD_AM57=1 ARM_LLVM_DIR=/opt/ti-llvm-3.6 DESTDIR=/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0/debian/ti-opencl returned exit code 2
make[1]: *** [debian/rules:25: override_dh_auto_build] Error 2
make[1]: Leaving directory '/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.0'
make: *** [debian/rules:9: build] Error 2
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
debuild: fatal error at line 1182:
dpkg-buildpackage -us -uc -ui failed
```

**Comment by RobertCNelson** [comment]
```find . -type f -exec sed -i 's:"CL/cl.h:"CL/TI/cl.h:g' {} \;```

**Comment by RobertCNelson** [comment]
```
In file included from /home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.1/host/src/api/api_platform.cpp:35:
/usr/include/CL/cl_ext.h:538:63: error: expected initializer before ‘CL_EXT_SUFFIX__VERSION_2_0_DEPRECATED’
                            size_t *     param_value_size_ret) CL_EXT_SUFFIX__VERSION_2_0_DEPRECATED;
                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/CL/cl_ext.h:548:82: error: expected initializer before ‘CL_EXT_SUFFIX__VERSION_2_0_DEPRECATED’
                                               size_t *     param_value_size_ret) CL_EXT_SUFFIX__VERSION_2_0_DEPRECATED;
                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[5]: *** [src/CMakeFiles/TIOpenCL.dir/build.make:115: src/CMakeFiles/TIOpenCL.dir/api/api_platform.cpp.o] Error 1
make[5]: Leaving directory '/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.1/build/am57'
make[4]: *** [CMakeFiles/Makefile2:265: src/CMakeFiles/TIOpenCL.dir/all] Error 2
make[4]: Leaving directory '/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.1/build/am57'
make[3]: *** [Makefile:152: all] Error 2
make[3]: Leaving directory '/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.1/build/am57'
make[2]: *** [makefile:93: install] Error 2
make[2]: Leaving directory '/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.1'
dh_auto_build: make -j1 BUILD_EVE_FIRMWARE=0 BUILD_OUTPUT=lib BUILD_AM57=1 ARM_LLVM_DIR=/opt/ti-llvm-3.6 DESTDIR=/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.1/debian/ti-opencl returned exit code 2
make[1]: *** [debian/rules:25: override_dh_auto_build] Error 2
make[1]: Leaving directory '/home/debian/repos/ti-opencl-10-debian/buster/ti-opencl_01.02.00.02-git20191223.1'
make: *** [debian/rules:9: build] Error 2
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
debuild: fatal error at line 1182:
dpkg-buildpackage -us -uc -ui failed
```

**Comment by RobertCNelson** [comment]
```debian@beaglebone:/usr/share/ti/examples/opencl/float_compute$ sudo ./float_compute 

This example computes y[i] = M[i] * x[i] + C on single precision floating point arrays of size 2097152
- Computation on the ARM is parallelized across the A15s using OpenMP.
- Computation on the DSP is performed by dispatching an OpenCL NDRange kernel across the compute units (C66x cores) in the compute device.

RunningTIOCL FATAL: Internal Error: Number of message queues (0) does not match number of compute units (2)
```

**Comment by RobertCNelson** [comment]
```192.168.1.4:/mnt/rcnee_pool/ti-processor-sdk/ /mnt/ti-processor-sdk/ nfs rw,nfsvers=3,rsize=8192,wsize=8192 0 1```

**Comment by RobertCNelson** [comment]
Okay, first 2 are done.. Stretch, v4.14.x-ti kernel, tidl 1.4.0, sdk 06.02.00.81..

https://rcn-ee.net/rootfs/bb.org/testing/2020-03-24/

**Comment by RobertCNelson** [comment]
Buster (10.x) userspace is now sdk 06.02.00.81, but only with v4.14.x-ti kernel, need to fix the v4.19.x-ti config..

**Comment by RobertCNelson** [comment]
Released as is, https://rcn-ee.net/rootfs/bb.org/testing/2020-03-26/
